x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/retriever/PinnedRetrieverBuilder.java

```diff
@@ -0,0 +1,271 @@
+        }
+    }
+
+    private void validateSort(SearchSourceBuilder source) {
+        List<SortBuilder<?>> sorts = source.sorts();
+        if (sorts == null || sorts.isEmpty()) {
+            return;
+        }
+        for (SortBuilder<?> sort : sorts) {
+            if (sort instanceof ScoreSortBuilder) {
+                continue;
+            }
+            if (sort instanceof FieldSortBuilder) {
+                FieldSortBuilder fieldSort = (FieldSortBuilder) sort;
+                if (ShardDocSortField.NAME.equals(fieldSort.getFieldName())) {
+                    continue;
+                }
+            }
+            throw new IllegalArgumentException(
+                "[" + NAME + "] retriever only supports sorting by score, invalid sort criterion: " + sort.toString()
+            );
+        }
+    }
+
+    public PinnedRetrieverBuilder(List<String> ids, List<SpecifiedDocument> docs, RetrieverBuilder retrieverBuilder, int rankWindowSize) {
+        super(new ArrayList<>(), rankWindowSize);
+        validateIdsAndDocs(ids, docs);
+        this.ids = ids;
+        this.docs = docs;
+        addChild(new PinnedRetrieverBuilderWrapper(retrieverBuilder));
+    }
+
+    public PinnedRetrieverBuilder(
+        List<String> ids,
+        List<SpecifiedDocument> docs,
+        List<RetrieverSource> retrieverSource,
+        int rankWindowSize,
+        String retrieverName,
+        List<QueryBuilder> preFilterQueryBuilders
+    ) {
+        super(retrieverSource, rankWindowSize);
+        validateIdsAndDocs(ids, docs);
+        this.ids = ids;
+        this.docs = docs;
+        this.retrieverName = retrieverName;
+        this.preFilterQueryBuilders = preFilterQueryBuilders;
+    }
+
+    @Override
+    public String getName() {
+        return NAME;
+    }
+
+    public int rankWindowSize() {
+        return rankWindowSize;
+    }
+
+    /**
+     * Creates a PinnedQueryBuilder with the appropriate pinned documents.
+     * Prioritizes docs over ids if both are present.
+     *
+     * @param baseQuery the base query to pin documents to
+     * @return a PinnedQueryBuilder
+     * @throws IllegalArgumentException if baseQuery is null
+     */
+    private QueryBuilder createPinnedQuery(QueryBuilder baseQuery) {
+        Objects.requireNonNull(baseQuery, "Underlying query cannot be null for pinned retriever");
+
+        if (docs != null && docs.isEmpty() == false) {
+            return new PinnedQueryBuilder(baseQuery, docs.toArray(new SpecifiedDocument[0]));
+        }
+        return new PinnedQueryBuilder(baseQuery, ids.toArray(new String[0]));
+    }
+
+    @Override
+    protected SearchSourceBuilder finalizeSourceBuilder(SearchSourceBuilder source) {
+        validateSort(source);
+        QueryBuilder underlyingQuery = source.query();
+        if (underlyingQuery == null) {
+            throw new IllegalArgumentException("[underlying query] must not be null for pinned retriever");
```

Comment:
We typically call out API objects that have meaning to users with square brackets, underlying query is something internal to our code. Suggest making this error more generic. 

I also have not validated if this check is required.

```suggestion
            throw new IllegalArgumentException("pinned retriever requires retriever with associated query");
```

Reply:
Had faced this error once in the ci pipeline. Thus introduced the check